### PR TITLE
[jax.numpy] Fix nanquantile on empty reduction axes

### LINKED ISSUE #39468: jnp.[nan]quantile & jnp.[nan]percentile fails on empty reduction axes
### The Problem

`jnp.nanquantile` raises an indexing or gather error when the
reduction axis is empty, however, NumPy returns `NaNs` with the expected
reduced shape.

### Reproduction

```python
import jax
import jax.numpy as jnp
import numpy as np

x_np = np.empty((0, 4), dtype=np.float32)
x_jax = jnp.asarray(x_np)

print(np.nanquantile(x_np, 0.5, axis=0))
print(jnp.nanquantile(x_jax, 0.5, axis=0))
print(jax.jit(lambda x: jnp.nanquantile(x, 0.5, axis=0))(x_jax))
```
NumPy returns:
```
[nan nan nan nan]
```

JAX instead raises an indexing / gather error.
but JAX should match:
```
np.nanquantile(np.empty((0, 4)), 0.5, axis=0)
```
and return `NaNs` with the appropriate reduced
shape for an empty reduction axis.

I also observed the failure with:
- `axis=None`
- tuple-valued axes
- ` keepdims=True`
- scalar and vector `q`
- eager execution and `jax.jit`

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.11.1.dev20260728+7238390c24
jaxlib: 0.11.0
numpy:  2.5.1
python: 3.12.12 (main, Oct 14 2025, 21:38:21) [Clang 20.1.4 ]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='Eds-MacBook-Pro.local', release='24.6.0', version='Darwin Kernel Version 24.6.0: Wed Nov  5 21:30:44 PST 2025; root:xnu-11417.140.69.705.2~1/RELEASE_ARM64_T6041', machine='arm64')
```

### PR DESCRIPTION
**Pull request by edluyuan**
## Summary

This PR aims to fixe `jnp.nanquantile` when the normalised reduction axis has size zero.

Before this, when JAX attempts to calculate quantile indices and gather values from an empty array, resulting in an indexing or gather error. This code change detects an empty reduction axis and returns a NaN-filled result with a shape matching the NumPy behaviour.

Also added Regression tests that cover scalar and vector quantiles and different axis configurations, `keepdims=True` and `keepdims=False`, eager execution, and compiled execution.

## The Problem

For an empty reduction axis, NumPy returns NaNs:

```python
import numpy as np

x = np.empty((0, 4), dtype=np.float32)
np.nanquantile(x, 0.5, axis=0)
# array([nan, nan, nan, nan], dtype=float32)
```

However, the equivalent JAX operation previously raised an indexing / gather error:

```python
import jax.numpy as jnp

x = jnp.empty((0, 4), dtype=jnp.float32)
jnp.nanquantile(x, 0.5, axis=0)
```

The issue also affects cases involving `axis=None`, tuple axes, vector-valued `q`, and `keepdims=True`.

## The Root Cause

When `_quantile` is called with `squash_nans=True`, it counts the number of valid, non-NaN elements along the reduction axis.

For an empty axis, this count is zero. However, the implementation continued to calculate the lower and upper quantile indices and then attempted to index into the empty sorted array.

## Implementations

The fix adds an early return when the normalized reduction axis is statically known to have size zero:

```python
core.definitely_equal(a_shape[axis], 0)
```

For this case, the implementation:

* 1. constructs the appropriate reduced output shape;
* 2. preserves the requested `keepdims` layout;
* 3. fills the result with NaNs using the input dtype; and
* 4.  returns before sorting and quantile-index calculation.

`core.definitely_equal` is used so that the check remains compatible with dimensions that may be symbolic rather than Python integers.

## Tests

A short regression test was added for:

* 1. testing one-dimensional and multidimensional empty arrays;
* 2. `axis=None`, integer axes, and tuple axes;
* 3. scalar and vector-valued `q`;
* 4. `keepdims=False` and `keepdims=True`;
* 5. eager execution
* 6. compiled execution.

Tests run:

```text
pytest -q tests/lax_numpy_reducers_test.py -k NanQuantileEmptyReduction
8 passed, 694 deselected, 32 subtests passed in 0.97s

pytest -q tests/lax_numpy_reducers_test.py -k WeightedQuantile
10 passed, 692 deselected in 0.90s

JAX_ENABLE_X64=True pytest -q tests/lax_numpy_reducers_test.py -k NanQuantileEmptyReduction
8 passed, 698 deselected, 32 subtests passed in 0.61s

pre-commit run --files jax/_src/numpy/reductions.py tests/lax_numpy_reducers_test.py
check python ast.........................................................Passed
check for merge conflicts................................................Passed
check toml...........................................(no files to check)Skipped
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
debug statements (python)................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
Pyrefly (type checking)..................................................Passed
jupytext.............................................(no files to check)Skipped
check copyright notice...................................................Passed
Lint GitHub Actions workflow files...................(no files to check)Skipped
```

Fixes #39468

### CODE REVIEW

**Comment by edluyuan** [comment]
Could a maintainer add the review label when convenient? Thanks!

**Comment by jakevdp** [comment]
It looks like this is also a problem for `quantile`, `percentile`, and `nanpercentile` as well. Since they all use the same underlying implementation, I think it would make sense to fix all at once.

**Comment by edluyuan** [comment]
> It looks like this is also a problem for `quantile`, `percentile`, and `nanpercentile` as well. Since they all use the same underlying implementation, I think it would make sense to fix all at once.

Hi Jake, thanks for pointing this out! 

I’ve added the fix in the shared `_quantile` implementation so it now fixes `quantile`, `nanquantile`, `percentile`, and `nanpercentile`.

Note: For a statically empty reduction axis, this version now returns NaNs with the reduced input shape before the sorting/ gather paths. so it preserves `axis=None`, tuple-valued axes, and `keepdims`, including NumPy’s empty-input shape behaviour for vector-valued `q`s.

I also expanded the regression tests a bit to cover all four APIs, scalar and vector q, eager execution, and JIT compilation:)
### Validation and Checks:
```
❯ pytest -q tests/lax_numpy_reducers_test.py \
  -k "QuantileEmptyReduction or WeightedQuantile"
..................                                                              [100%]
18 passed, 684 deselected, 128 subtests passed in 2.24s
❯ JAX_ENABLE_X64=True pytest -q tests/lax_numpy_reducers_test.py \
  -k "QuantileEmptyReduction or WeightedQuantile"
..................                                                              [100%]
18 passed, 688 deselected, 128 subtests passed in 2.29s
❯ pre-commit run --files \
  jax/_src/numpy/reductions.py \
  tests/lax_numpy_reducers_test.py
check python ast.........................................................Passed
check for merge conflicts................................................Passed
check toml...........................................(no files to check)Skipped
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
debug statements (python)................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
Pyrefly (type checking)..................................................Passed
jupytext.............................................(no files to check)Skipped
check copyright notice...................................................Passed
Lint GitHub Actions workflow files...................(no files to check)Skipped
```

File: tests/lax_numpy_reducers_test.py [hunk 1]
```diff
@@ -1118,5 +1118,47 @@ def jnp_fun(data, q, weights):
...
+    ((0, 4), 0),
+    ((0, 4), None),
+    ((3, 0), 1),
+    ((3, 0), None),
+    ((2, 0, 3), (0, 1)),
+    ((2, 0, 3), (1, 2)),
+  )
+  @jtu.ignore_warning(category=RuntimeWarning, message="Mean of empty slice")
+  def testQuantileEmptyReduction(self, shape, axis):
+    for opname in [
```
Comment (jakevdp) [inline tests/lax_numpy_reducers_test.py]: opname should be part of the test parameterization, not a for loop within the test (you can follow the approach used in `testQuantile` above)

File: tests/lax_numpy_reducers_test.py [hunk 2]
```diff
@@ -1118,5 +1118,47 @@ def jnp_fun(data, q, weights):
...
+      is_percentile = "percentile" in opname
+      q_values = [
+        50.0 if is_percentile else 0.5,
+        np.array(
+          [25.0, 50.0, 75.0] if is_percentile
+          else [0.25, 0.5, 0.75],
+          dtype=np.float32),
+      ]
+
+      for q in q_values:
```
Comment (jakevdp) [inline tests/lax_numpy_reducers_test.py]: These functions are vectorized, so we should pass `q_values` as a vector rather than iterating over it.

File: tests/lax_numpy_reducers_test.py [hunk 3]
```diff
@@ -1118,5 +1118,47 @@ def jnp_fun(data, q, weights):
...
+        for keepdims in [False, True]:
```
Comment (jakevdp) [inline tests/lax_numpy_reducers_test.py]: `keepdims` should be part of the test parameterization rather than iterated in a `for` loop.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/lax_numpy_reducers_test.py (hunk 3 above)
Comment (edluyuan) [reply tests/lax_numpy_reducers_test.py]: Thanks! these suggestions make sense. I’ve updated the regression test with:

- 1. `opname` is parameterized through `jtu.sample_product`;
- 2. `keepdims` is also part of the test parameterisation
- 3. the quantiles are passed as a single vector rather than iterated.

removed the nested loops and gives each API, shape/axis pair, and keepdims value independently reported test case. 

The eager and compiled checks are still both covered.


### Checks:
```
❯ pytest -q tests/lax_numpy_reducers_test.py \
  -k "QuantileEmptyReduction or WeightedQuantile"
....................                                                                                                                                                                                            [100%]
20 passed, 684 deselected in 1.51s
❯ JAX_ENABLE_X64=True pytest -q tests/lax_numpy_reducers_test.py \
  -k "QuantileEmptyReduction or WeightedQuantile"
....................                                                                                                                                                                                            [100%]
20 passed, 688 deselected in 1.08s
❯ pre-commit run --files \
  jax/_src/numpy/reductions.py \
  tests/lax_numpy_reducers_test.py
check python ast.........................................................Passed
check for merge conflicts................................................Passed
check toml...........................................(no files to check)Skipped
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
debug statements (python)................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
Pyrefly (type checking)..................................................Passed
jupytext.............................................(no files to check)Skipped
check copyright notice...................................................Passed
Lint GitHub Actions workflow files...................(no files to check)Skipped
```

**Comment by edluyuan** [comment]
`opname` and `keepdims` are now test parameters, and q is passed as a vector. Thanks!

File: tests/lax_numpy_reducers_test.py [hunk 4]
```diff
@@ -1118,47 +1118,39 @@ def jnp_fun(data, q, weights):
...
+  def testQuantileEmptyReduction(self, shape, axis, opname, keepdims):
+    is_percentile = "percentile" in opname
+    q = np.array(
+      [25.0, 50.0, 75.0] if is_percentile else [0.25, 0.5, 0.75],
+      dtype=np.float32,
+    )
+    args_maker = lambda: [np.empty(shape, dtype=np.float32)]
+
+    # NumPy uses the reduced nanmean shape for empty axes, even for vector q.
+    np_fun = partial(np.nanmean, axis=axis, keepdims=keepdims)
```
Comment (jakevdp) [inline tests/lax_numpy_reducers_test.py]: We should test directly against the associated NumPy operation to ensure we're matching NumPy's behavior. We don't want our tests to depend on NumPy implementation details that may change without warning in the future.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/lax_numpy_reducers_test.py (hunk 4 above)
Comment (edluyuan) [reply tests/lax_numpy_reducers_test.py]: Sure!. I’ve updated the test to compare directly against `np.nanquantile` and `np.nanpercentile` instead of using `np.nanmean` as oracle.

I've also done some digging in numpy and found that NumPy `2.5.1 ` only has an explicit empty-array fallback in the NaN-skipping quantile implementation; that path returns `np.nanmean(...)`. The regular quantile implementation has no corresponding empty fallback.

Therefore the correct behaviour and test should be:
- `nanquantile` and `nanpercentile` should return NaNs and be compared directly against NumPy
- `quantile` and `percentile` should not return NaNs; they should retain error behaviour. 


Again, The regression test parameterises the operation and `keepdims`, passes `q` as a vector,/ covers `axis=None`, integer axes, tuple axes, eager execution, and jitted execution.
###
```
❯ pytest -q tests/lax_numpy_reducers_test.py \
  -k "NanQuantileEmptyReduction or WeightedQuantile"

JAX_ENABLE_X64=True pytest -q tests/lax_numpy_reducers_test.py \
  -k "NanQuantileEmptyReduction or WeightedQuantile"

pre-commit run --files \
  jax/_src/numpy/reductions.py \
  tests/lax_numpy_reducers_test.py
....................                                                                                                                                                                                            [100%]
20 passed, 684 deselected in 1.49s
....................                                                                                                                                                                                            [100%]
20 passed, 688 deselected in 1.05s
check python ast.........................................................Passed
check for merge conflicts................................................Passed
check toml...........................................(no files to check)Skipped
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
debug statements (python)................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
Pyrefly (type checking)..................................................Passed
jupytext.............................................(no files to check)Skipped
check copyright notice...................................................Passed
Lint GitHub Actions workflow files...................(no files to check)Skipped
```

**Comment by edluyuan** [comment]
updated the test to compare directly against `np.nanquantile `and `np.nanpercentile`

**Comment by jakevdp** [comment]
Can you please squash your changes into a single commit?

**Comment by edluyuan** [comment]
> Can you please squash your changes into a single commit?

Done!

File: jax/_src/numpy/reductions.py [hunk 5]
```diff
@@ -2572,6 +2572,19 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
...
+  if squash_nans and core.definitely_equal(a_shape[axis], 0):
+    # For an empty reduction axis, return NaNs with the reduced input shape.
+    result_shape = list(a_shape)
+    if keepdims:
+      result_shape[axis] = 1
+    else:
+      result_shape.pop(axis)
+    result = lax.full(tuple(result_shape), np.nan, dtype=a.dtype)
+    if keepdims and keepdim:
+      result = result.reshape(keepdim)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: Rather than creating `result` and reshaping to `keepdim`, we should directly use `result_shape = keepdim` above.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/reductions.py (hunk 5 above)
Comment (edluyuan) [reply jax/_src/numpy/reductions.py]: Done,  the empty-axis path uses `keepdim` directly when it already represents the final `keepdims=True` output shape, no longer allocating an intermediate result and reshaping it.

File: jax/_src/numpy/reductions.py [hunk 6]
```diff
@@ -2572,6 +2572,19 @@ def _quantile(a: Array, q: Array, axis: int | tuple[int, ...] | None,
 
   a_shape = a.shape
   q_orig = q
+
+  if squash_nans and core.definitely_equal(a_shape[axis], 0):
+    # For an empty reduction axis, return NaNs with the reduced input shape.
+    if keepdims and keepdim:
+      result_shape = keepdim
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py:2580]: Is this always the correct shape? In the non-empty path, `keepdim` is modified if `q.ndim > 0`: https://github.com/jax-ml/jax/blob/89c171583f0002e793aea5b46691542e58638ac8/jax/_src/numpy/reductions.py#L2708-L2710

Comment (edluyuan) [reply jax/_src/numpy/reductions.py:2580]: Yes this is intentional for the empty-input path. for NumPy 2.5.1 special-cases `a.size == 0 `by returning `np.nanmean(a, axis, keepdims=keepdims)`, so unlike the normal quantile path, vector-valued `q` does not prepend a `q `dimension. The regression test uses vector `q `and compares directly against NumPy, so this shape behavior is covered. I’ve added a comment to make the distinction explicit.

Btw, apologies for the delayed response! I had some travel over the past few weeks and was a bit slower than usual in getting back to this. Thanks for your patience!

**Review by jakevdp — APPROVED** [review]
Thanks!
